src/wasm-type.h

```diff
@@ -23,16 +23,23 @@
 
 namespace wasm {
 
+class Type;
+struct Tuple;
+struct Signature;
+struct Struct;
+struct Array;
+
+typedef std::vector<Type> TypeList;
+
 class Type {
   // The `id` uniquely represents each type, so type equality is just a
-  // comparison of the ids. For basic types the `id` is just the `ValueType`
+  // comparison of the ids. For basic types the `id` is just the `ID`
   // enum value below, and for constructed types the `id` is the address of the
   // canonical representation of the type, making lookups cheap for all types.
   uintptr_t id;
-  void init(const std::vector<Type>&);
 
 public:
-  enum ValueType : uint32_t {
+  enum ID : uint32_t {
```

Comment:
Given that values here are IDs of all the types incl. reference types and tuples, and that there is also the notion of `HeapType` and `StorageType` in the spec, I renamed the enum to `ID` for clarity. That ok?

Reply:
Sounds good! The old name made a lot more sense before we decided to make it an internal detail.

Reply:
My mental model so far is

| Type | Basic | Value | Single | Reference | Constructed
|------|-------|-------|------|-------|----
| none | x |
| unreachable | x | ? | x
| i32 | x | x | x
| i64 | x | x | x
| f32 | x | x | x
| f64 | x | x | x
| v128 | x | x | x
| funcref | x | | x | x
| externref | x | | x | x
| nullref | x | | x | x
| exnref | x | | x | ?
| Tuple | ? | ? | | | x
| Signature | | | | x | x
| Struct | | | | x | x
| Array | | | | x | x

so the closest name is indeed `BasicID`. Depending on how `basic` tuples are, `NonConstructedID` would be even more concise but long. I guess I named it just `ID` because it's more ergonomic and doesn't require thinking about the differences too much (what can have a static `ID` simply has a static `ID`). Ultimately it doesn't really matter how we name it I guess, because we typically just write `Type::something` anyway. Remains the thought about it being a little less ergonomic / a little less future-proof maybe, but if you'd still prefer `BasicID` after my tldr here I'm fine with renaming it, just let me know :)

Reply:
Going to rename to `BasicID`. Figured that it's used at so few places by name that it really doesn't matter much.

Reply:
Nice chart! What does the `Value` column mean, though? If `Value` just means `isSingle() && !isRef()`, then I don't think we need to give it its own name. I've been thinking of `Basic` types / `ValueType`s as being any type that is not parameterized or constructed, i.e. any type that has a static ID. So `Type::none` and `Type::unreachable` would be `Basic`, but tuples would not be `Basic`. Also, I don't think `Type::unreachable` should be `Single` because it can take the place of any type, including tuple types.

Reply:
Per the spec, [value types](https://webassembly.github.io/spec/core/syntax/types.html#value-types) are `i32`, `i64`, `f32` and `f64` as of the MVP, and `v128` by extension. My understanding is that these are the types whose values are always passed "by value" and can be trivially stored and loaded in memory. As such, these are `isSingle() && !isTuple() && !isRef()` (when considering `unreachable` not single), and we might need to know if a type `isValue()` on trivial `Load`s and `Store`s.